src/LootTable.h

```diff
@@ -0,0 +1,620 @@
+		bool a_Expand,
+		int a_Weight,
+		int a_Quality
+	):
+		m_Conditions(std::move(a_Conditions)),
+		m_Functions(std::move(a_Functions)),
+		m_Type(a_Type),
+		m_Content(std::move(a_Name)),
+		m_Expand(a_Expand),
+		m_Weight(a_Weight),
+		m_Quality(a_Quality)
+	{
+	}
+
+
+
+
+
+	cLootTablePoolEntry(
+		cLootTableConditionVector a_Conditions,
+		cLootTableFunctionVector a_Functions,
+		enum LootTable::ePoolEntryType a_Type,
+		cLootTablePoolEntryVector a_Children,
+		int a_Weight,
+		int a_Quality
+	):
+		m_Conditions(std::move(a_Conditions)),
+		m_Functions(std::move(a_Functions)),
+		m_Type(a_Type),
+		m_Content(std::move(a_Children)),
+		m_Weight(a_Weight),
+		m_Quality(a_Quality)
+	{
+	}
+
+	cLootTableConditionVector m_Conditions;
+	cLootTableFunctionVector m_Functions;
+	enum LootTable::ePoolEntryType m_Type;
+	std::variant<cItem, AString, cLootTablePoolEntryVector> m_Content;
+	bool m_Expand;
+	int m_Weight;
+	int m_Quality;
+};
+
+
+
+
+/** Represents a while loot table pool */
+struct cLootTablePool;
+typedef std::vector<cLootTablePool> cLootTablePoolVector;
+struct cLootTablePool
+{
+	cLootTablePool(
+		cLootTablePoolRolls a_Rolls,
+		cLootTablePoolEntryVector a_Entries,
+		cLootTableConditionVector a_Conditions
+	):
+		m_Rolls(a_Rolls),
+		m_Entries(std::move(a_Entries)),
+		m_Conditions(std::move(a_Conditions))
+	{
+		for (const auto & Entry : m_Entries)
+		{
+			m_TotalWeight += Entry.m_Weight;
+			m_TotalQuality += Entry.m_Weight;
+		}
+	}
+
+
+
+
+
+	cLootTablePool(
+		cLootTablePoolRolls a_Rolls,
+		cLootTablePoolRolls a_BonusRolls,
+		cLootTablePoolEntryVector a_Entries,
+		cLootTableConditionVector a_Conditions
+	):
+		m_Rolls(a_Rolls),
+		m_BonusRolls(a_BonusRolls),
+		m_Entries(std::move(a_Entries)),
+		m_Conditions(std::move(a_Conditions))
+	{
+		for (const auto & Entry : m_Entries)
+		{
+			m_TotalWeight += Entry.m_Weight;
+			m_TotalQuality += Entry.m_Weight;
+		}
+	}
+
+	cLootTablePoolRolls m_Rolls;
+	cLootTablePoolRolls m_BonusRolls;
+	cLootTablePoolEntryVector m_Entries;
+	cLootTableConditionVector m_Conditions;
+	int m_TotalWeight = 0;
+	int m_TotalQuality = 0;
+};
+
+
+
+
+
+struct EnumClassHash
+{
+	template <typename T>
+	std::size_t operator()(T t) const
+	{
+		return static_cast<std::size_t>(t);
+	}
+};
+
+
+
+
+/** A individual loot table */
+class cLootTable
+{
+public:
+	/** Creates a empty loot table with type empty */
+	cLootTable();
+
+	/** Creates new loot table from string describing the loot table */
+	cLootTable(const Json::Value & a_Description, cWorld * a_World);
+
+	cLootTable(const cLootTable & a_Other) = default;
+	cLootTable(cLootTable && a_Other) = default;
+	cLootTable & operator = (const cLootTable & a_Other) = default;
+	cLootTable & operator = (cLootTable && a_Other) = default;
+
+	/** Fills the specified block entity at the position with loot and returns the success */
+	bool FillWithLoot(cItemGrid & a_ItemGrid, cWorld * a_World, const Vector3i & a_Pos, const UInt32 & a_Player) const;
```

Comment:
is there a reason to remove the const qualifyer?

I'm just curious

Reply:
And the block entity recieves a pointer to the world so I did the same here

Reply:
The `const` has no meaning in a declaration when passing by-value (a copy is made anyway, so the original is not modified). 

As for by-pointer vs by-reference vs by-value:
Stuff passed by-pointer means that it may be optional (`nullptr`). When you pass by-reference, you maek it clear to the caller that the parameter must be valid. Filling with loot requires a valid world, so we make it clear by passing by-reference. By-value passing is used for tiny stuff that is cheap to copy (can fit into a few registers) - basically plain numbers and `Vector3<>` instances were decided to be small enough to always pass by-value.

In some old parts of the code, pointers are still used instead of references, but for new code we should try to make the optional/required distinction.

Reply:
That makes sense

Oh ok - yeah the optional stuff is a bit weird. Block entities don't get a pointer to their world when loaded from disk - i tried accesing the world in the constructor and kept getting acces violations (early trying phase).

This makes sense - did apply those changes to all methods